<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-основной-информации-о-датафрейме" data-toc-modified-id="Изучение-основной-информации-о-датафрейме-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение основной информации о датафрейме</a></span><ul class="toc-item"><li><span><a href="#Импорт-необходимых-библиотек" data-toc-modified-id="Импорт-необходимых-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт необходимых библиотек</a></span></li><li><span><a href="#Открытие-файла" data-toc-modified-id="Открытие-файла-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Открытие файла</a></span></li><li><span><a href="#Вывод-первых-десяти-строк-таблицы" data-toc-modified-id="Вывод-первых-десяти-строк-таблицы-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод первых десяти строк таблицы</a></span></li><li><span><a href="#Вывод-основной-информации-о-датафрейме" data-toc-modified-id="Вывод-основной-информации-о-датафрейме-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод основной информации о датафрейме</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Разбиение-данных-на-выборки" data-toc-modified-id="Разбиение-данных-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбиение данных на выборки</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-моделей" data-toc-modified-id="Исследование-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование моделей</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверка-модели-на-тестовой-выборке" data-toc-modified-id="Проверка-модели-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка модели на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверка-модели-на-адекватность" data-toc-modified-id="Проверка-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка модели на адекватность</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Рекомендация тарифов

В распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф.

**Цель исследования** — построить лучшую модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75.

**Ход исследования**
1. Изучение файла
2. Разбиение данных на выборки
3. Исследование моделей классификации
4. Проверка модели на тестовой выборке
5. Проверка модели на адекватность
6. Вывод

## Изучение основной информации о датафрейме

### Импорт необходимых библиотек

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split 

### Открытие файла

In [2]:
try:
    data = pd.read_csv('/datasets/users_behavior.csv', sep=',', engine='python')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/users_behavior.csv', sep=',', engine='python')

### Вывод первых десяти строк таблицы

In [3]:
data.head(10)

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0

### Вывод основной информации о датафрейме

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


### Вывод

Согласно документации к данным:
* `сalls` — количество звонков,
* `minutes` — суммарная длительность звонков в минутах,
* `messages` — количество sms-сообщений,
* `mb_used` — израсходованный интернет-трафик в Мб,
* `is_ultra` — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

Целевым признаком будет являться `is_ultra`, а все остальные столбцы — признаками.

## Разбиение данных на выборки

Делим исходные данные на тренировочную, валидационную и тестовую выборки в отношении 3:1:1.

In [5]:
data_train, data_sep = train_test_split(data, test_size=0.4, random_state=12345)
data_valid, data_test = train_test_split(data_sep, test_size=0.5, random_state=12345)

Из каждой выборки берем признаки и целевой признак.

In [6]:
features_train = data_train.drop(['is_ultra'], axis=1) #для тренировочной выборки
target_train = data_train['is_ultra']

features_valid = data_valid.drop(['is_ultra'], axis=1) #для валидационной выборки
target_valid = data_valid['is_ultra']

features_test = data_test.drop(['is_ultra'], axis=1) #для тестовой выборки
target_test = data_test['is_ultra']

In [7]:
data_train.shape

(1928, 5)

In [8]:
data_valid.shape

(643, 5)

In [9]:
data_test.shape

(643, 5)

### Вывод

Данные были разделены на тренировочную, валидационную и тестовую в отношении 3:1:1: в тренировочной выборке 1928 объектов, а в валидационной и тестовой по 643 объекта.

## Исследование моделей

### Дерево решений

Возьмем для модели дерева решений значения максимальной глубины от 1 до 10 и выберем лучшую модель по значению accuracy.

In [10]:
best_model_tree = None
best_result_tree = 0
best_depth = 0

for depth in range(1, 11):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model_tree.fit(features_train, target_train)
    predictions_valid = model_tree.predict(features_valid)
    result = accuracy_score(target_valid, predictions_valid) 
    
    if result > best_result_tree:
        best_model_tree = model_tree
        best_result_tree = result
        best_depth = depth
        
print('Accuracy наилучшей модели дерева решений на валидационной выборке равно', best_result_tree, 
      'при глубине дерева', best_depth)

Accuracy наилучшей модели дерева решений на валидационной выборке равно 0.7853810264385692 при глубине дерева 3


### Случайный лес

Возьмем для модели случайного леса значения максимальной глубины от 1 до 10 и выберем лучшую модель по значению accuracy.

In [11]:
best_model_forest = None
best_result_forest = 0
best_est = 0

for est in range(1, 11):
    model_forest = RandomForestClassifier(random_state=12345, n_estimators=est)
    model_forest.fit(features_train, target_train) 
    predictions_valid = model_forest.predict(features_valid)
    result = accuracy_score(target_valid, predictions_valid) 
    
    if result > best_result_forest:
        best_model_forest = model_forest
        best_result_forest = result
        best_est = est

print('Accuracy наилучшей модели случайного леса на валидационной выборке равно', best_result_forest, 
      'с количеством деревьев, равным', best_est)

Accuracy наилучшей модели случайного леса на валидационной выборке равно 0.7853810264385692 с количеством деревьев, равным 10


In [12]:
best_model_forest = None
best_result_forest = 0
best_est = 0

for est in range(10, 201, 10):
    model_forest = RandomForestClassifier(random_state=12345, n_estimators=est)
    model_forest.fit(features_train, target_train) 
    predictions_valid = model_forest.predict(features_valid)
    result = accuracy_score(target_valid, predictions_valid) 
    
    if result > best_result_forest:
        best_model_forest = model_forest
        best_result_forest = result
        best_est = est

print('Accuracy наилучшей модели случайного леса на валидационной выборке равно', best_result_forest, 
      'с количеством деревьев, равным', best_est)

Accuracy наилучшей модели случайного леса на валидационной выборке равно 0.7916018662519441 с количеством деревьев, равным 50


### Логистическая регрессия

In [13]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train, target_train)

predictions_valid = model.predict(features_valid)

result = accuracy_score(target_valid, predictions_valid) 

print('Accuracy модели логистической регрессии на валидационной выборке:', result)

Accuracy модели логистической регрессии на валидационной выборке: 0.7107309486780715


### Вывод

У модели логистической регрессии недостаточное значение accuracy на валидационной выборке (0.71), а у моделей дерева решений и случайного леса значение accuracy одинаковое (0.785). Учитывая, что модель случайного леса в целом точнее, чем модель дерева решений, то стоит взять модель случайного леса как лучшую.

## Проверка модели на тестовой выборке

In [14]:
predictions_test = best_model_forest.predict(features_test)
result = accuracy_score(target_test, predictions_test) 
result

0.7931570762052877

### Вывод

Accuracy выбранной модели случайного леса на тестовой выборке равно 0.79, что составило больше требуемых 0.75, значит, модель выбрана верно.

## Проверка модели на адекватность

Возьмем три случайные модели и посчитаем их значение accuracy по тестовой выборке:
1. Модель предсказывает ответы с вероятностью 50/50;
2. Модель предсказывает целевой признак всегда как 1;
3. Модель предсказывает целевой признак всегда как 0

Для первой модели значение accuracy будет равно 0.5. Для остальных моделей значение необходимо посчитать, так как для нашего набора данных классы распределены не поровну.

Посчитаем значение accuracy для модели, предсказывающей целевой признак всегда как 1. Количество правильных ответов будет равно количеству значений целевого признака, равных 1.

In [15]:
#количество значений целевого признака, равных 1
amount_1 = data_test.loc[data_test['is_ultra'] == 1].shape[0]
#количество всех объектов
all_amount = data_test.shape[0]

accuracy_1 = amount_1 / all_amount

accuracy_1

0.3157076205287714

Теперь посчитаем значение accuracy для модели, предсказывающей целевой признак всегда как 0. Количество правильных ответов будет равно количеству значений целевого признака, равных 0.

In [16]:
#количество значений целевого признака, равных 0
amount_0 = data_test.loc[data_test['is_ultra'] == 0].shape[0]

accuracy_0 = amount_0 / all_amount

accuracy_0

0.6842923794712286

### Вывод

Значение accuracy выбранной лучшей модели равно 0.79, что оказалось больше, чем значения случайных моделей. Таким образом, выбранная модель прошла проверку на адекватность.

## Вывод

Были рассмотрены модели случайного леса, дерева решений и логистической регрессии. Для моделей случайного леса и дерева решений были перебраны гиперпараметры глубины и количества деревьев соответственно от 1 до 10. Из данных моделей была выбрана модель случайного дерева, так как она имеет наибольшее значение accuracy, чем у логистической регрессии, а также в целом большую точность, чем у дерева решений.

Проверка модели на тестовой выборке показало значение accuracy большее, чем требуемое 0.75, поэтому модель проверку прошла.

Также модель прошла проверку на адекватность: она показала наибольшее значение accuracy, чем случайные модели, которые предсказывают ответы, равные полностью одному из классов, либо же модель, предсказывающая с вероятностью 50/50.